### PR TITLE
Feat(Array): Add `arrayMinIndex` and `arrayMaxIndex`

### PR DESCRIPTION
**Pull request by fallintoplace**
### Changelog category (leave one):
- New Feature

### Changelog entry (a [user-readable short description](https://github.com/ClickHouse/ClickHouse/blob/master/docs/changelog_entry_guidelines.md) of the changes that goes into CHANGELOG.md):
Add `arrayMinIndex` and `arrayMaxIndex` array functions.

### Description

**What**
- Add `arrayMinIndex` and `arrayMaxIndex`.
- Return the 1-based position of the first minimum or maximum element.
- Return `0` for an empty array.
- Support the same lambda and multiple-array forms as [`arrayMin`](https://clickhouse.com/docs/sql-reference/functions/array-functions#arraymin) and [`arrayMax`](https://clickhouse.com/docs/sql-reference/functions/array-functions#arraymax).

Examples:
```sql
arrayMaxIndex([1, 7, 3, 7]) -- 2
arrayMinIndex(x -> abs(x), [-10, 4, 8]) -- 2
arrayMaxIndex((x, y) -> x * y, [1, 5, 3], [2, 2, 5]) -- 3
```

**Why**
- This makes the position of an array extreme available directly.
- It avoids awkward extrema plus index lookup expressions.
- It follows the existing [`arrayFirst`](https://clickhouse.com/docs/sql-reference/functions/array-functions#arrayfirst), [`arrayFirstIndex`](https://clickhouse.com/docs/sql-reference/functions/array-functions#arrayfirstindex), [`arrayLast`](https://clickhouse.com/docs/sql-reference/functions/array-functions#arraylast), and [`arrayLastIndex`](https://clickhouse.com/docs/sql-reference/functions/array-functions#arraylastindex) conventions.
- Equal extrema keep the first position, matching [`arrayMin`](https://clickhouse.com/docs/sql-reference/functions/array-functions#arraymin) and [`arrayMax`](https://clickhouse.com/docs/sql-reference/functions/array-functions#arraymax).
- Existing functions are unchanged.

**Existing usage**
- The exact composite `indexOf(arr, arrayMin(arr))` appears in a public [ClickHouse Bloom filter example](https://gist.github.com/ardenwick/74e3d51ef08dafe4f437e7eaa99a1a4d).
- That pattern first computes the extreme value and then scans the same array again to recover its position. `arrayMinIndex(arr)` expresses the intended result directly.
- The exact composite `indexOf(hourly_values, arrayMax(hourly_values))` appears in a [ClickHouse monitoring example](https://oneuptime.com/blog/post/2026-03-31-clickhouse-arraymin-arraymax/view).
- In that case, the returned index identifies the hour containing the peak value. The same shape is useful for per-row scores, bucketed measurements, and other arrays where the value's position matters.
- The official [ClickHouse Advent of Code 2025 Day 3 article](https://clickhouse.com/blog/clickhouse-advent-of-code-2025?utm_source=chatgpt.com) finds the best digit in a slice with `arrayFirstIndex(x -> x = arrayMax(...), ...)`. `arrayMaxIndex(...)` expresses the same lookup directly.

**Other libraries**
- [NumPy](https://numpy.org/doc/stable/reference/generated/numpy.argmin.html) provides `argmin` and [argmax](https://numpy.org/doc/stable/reference/generated/numpy.argmax.html), with axis support and first-occurrence tie behavior. Its indices are zero-based.
- [PyTorch](https://docs.pytorch.org/docs/stable/generated/torch.argmax.html) and [JAX](https://docs.jax.dev/en/latest/_autosummary/jax.numpy.argmax.html) provide tensor reductions with dimension or axis selection.
- [TensorFlow](https://www.tensorflow.org/api_docs/python/tf/math/argmax) provides `tf.math.argmax` with axis and output-type options.
- [pandas](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.idxmax.html) provides `idxmax` and `idxmin`, but returns row or column labels instead of integer positions and skips missing values by default.
- [Julia](https://docs.julialang.org/en/v1/base/arrays/#Base.argmin) provides `argmin` and `argmax` for arrays and returns an array index.
- ClickHouse keeps its existing 1-based array position convention, while these APIs generally expose zero-based positions or labels.

**Related ClickHouse PRs**
- [#110163](https://github.com/ClickHouse/ClickHouse/pull/110163) vectorized numeric `arrayMin` and `arrayMax`.
- [#33415](https://github.com/ClickHouse/ClickHouse/pull/33415) added `arrayLast`.
- [#33465](https://github.com/ClickHouse/ClickHouse/pull/33465) added `arrayLastIndex`.
- [#81197](https://github.com/ClickHouse/ClickHouse/pull/81197) updated the `arrayFirst`, `arrayFirstIndex`, `arrayLast`, and `arrayLastIndex` family.

**Local benchmark**

Standalone benchmarks on AMD EPYC 9V74 and ARM64 show performance roughly comparable to `indexOf(arrayMin(...)) + indexOf(arrayMax(...))`, with results varying by type and array length.

**Implementation**
- Reuse `FunctionArrayMapped` for lambda and multiple-array support.
- Reuse existing `findExtremeMin/Max` for numeric reductions, followed by a forward scan for first-tie semantics.
- Use a one-pass scan for short arrays and a generic `compareAt` fallback for unsupported types.
- Preserve existing first-tie, special-value, empty-array, nullable, decimal, date, and wide-integer behavior.
- Leave existing functions unchanged.

---
Workflow [[PR](https://s3.amazonaws.com/clickhouse-test-reports/praktika.html?PR=120097&sha=latest&name_0=PR)]
Sync PR [[sync-upstream/pr/120097](https://github.com/search?q=head%3Async-upstream%2Fpr%2F120097+org%3AClickHouse+type%3Apr&type=pullrequests)]

### CODE REVIEW

**Comment by fallintoplace** [comment]
@alexey-milovidov What do you think about adding these functions?

File: src/Functions/array/arrayMinMaxIndex.cpp [hunk 1]
```diff
@@ -0,0 +1,1395 @@
...
+{
+
+enum class ArrayMinMaxIndexStrategy : uint8_t
+{
+    Min,
+    Max,
+};
+
+#if USE_MULTITARGET_CODE
+DECLARE_X86_64_V3_SPECIFIC_CODE(
```
Comment (Algunenano) [inline src/Functions/array/arrayMinMaxIndex.cpp]: This is wrong. We don't need DECLARE_X86_64_V3_SPECIFIC_CODE for USE_MULTITARGET_CODE. Just check for `__AVX2__` which is the default

File: src/Functions/array/arrayMinMaxIndex.cpp [hunk 2]
```diff
@@ -0,0 +1,1395 @@
...
+            return i;
+    }
+
+    const auto findInVector = [&](size_t offset)
+    {
+        unsigned mask;
+
+        if constexpr (std::is_same_v<T, Float32>)
+        {
+            const auto values = _mm256_loadu_ps(data + offset);
```
Comment (Algunenano) [inline src/Functions/array/arrayMinMaxIndex.cpp]: Please avoid intrinsics. This function does not need it and it requires writing the same thing 3 times for a new function nobody uses.

BTW, check findExtremeMinIndex instead, which already implements this. This code is basically ColumnVector::getPermutation limit 1 with extra complexity.

File: tests/performance/array_min_max_index.xml [hunk 3]
```diff
@@ -0,0 +1,679 @@
...
+        representative raw, lambda, multi-array, floating point, String, and Nullable cases.
+    -->
+    <tag>long</tag>
+
+    <settings>
+        <max_threads>1</max_threads>
+        <max_insert_threads>1</max_insert_threads>
+    </settings>
+
+    <substitutions>
```
Comment (Algunenano) [inline tests/performance/array_min_max_index.xml]: I guarantee nobody has, or even will, read this file as is. Having so many queries and replacements is a waste of CI resources. Please keep this small and simple

**Review by Algunenano — CHANGES_REQUESTED** [review]

File: tests/performance/array_min_max_index.xml (hunk 3 above)
Comment (fallintoplace) [reply tests/performance/array_min_max_index.xml]: Could you please rerun CI? I just simplified the benchmark. There seem to be some CI issues currently preventing the tests from running.

File: src/Functions/array/arrayMinMaxIndex.cpp (hunk 2 above)
Comment (fallintoplace) [reply src/Functions/array/arrayMinMaxIndex.cpp]: I’m aware of findExtremeMinIndex, but I don’t think it matches the required semantics. For example, with [5, 3, 3, 7], it can return the last 3, while arrayMinIndex should return the first one.

The custom indexed SIMD paths are what make this fast and make the feature worth having.

Comment (Algunenano) [reply src/Functions/array/arrayMinMaxIndex.cpp]: Please measure it and consider ways of reusing the existing code and optimizations.

I do not like the idea of adding, testing, reviewing and maintaining hand written intrinsics for a feature nobody asked for and that doesn't cover NEON (which as of today is the most common architecture).

The first goal should be simple code, and only make it complex when the extra performance is needed and can't be handled any other way (like dynamic dispatch, compiler instructions/flags/etc)

Comment (fallintoplace) [reply src/Functions/array/arrayMinMaxIndex.cpp]: I simplified this and removed the handwritten AVX paths. It now reuses findExtremeMin/Max plus a forward scan for first-tie semantics.

Benchmark results are roughly comparable to the composite on both tested machines.

Comment (fallintoplace) [reply src/Functions/array/arrayMinMaxIndex.cpp]: Existing usage

The exact composite indexOf(arr, arrayMin(arr)) appears in a public [ClickHouse Bloom filter example](https://gist.github.com/ardenwick/74e3d51ef08dafe4f437e7eaa99a1a4d).
The exact composite indexOf(hourly_values, arrayMax(hourly_values)) appears in a [ClickHouse monitoring example](https://oneuptime.com/blog/post/2026-03-31-clickhouse-arraymin-arraymax/view).

I have added this section as well to the description.

File: src/Functions/array/arrayMinMaxIndex.cpp [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+
+    for (size_t i = 0; i < size; ++i)
+        if (data[i] == extreme)
+            return i;
+    return size;
+}
+
+template <ArrayMinMaxIndexStrategy strategy, typename T>
+requires(has_find_extreme_implementation<T> || underlying_has_find_extreme_implementation<T>)
+static size_t findIndexUsingExtreme(const T * data, size_t size)
```
Comment (Algunenano) [inline src/Functions/array/arrayMinMaxIndex.cpp]: Isn't this what `findExtremeMinIndex` / `findExtremeMaxIndex` do?

File: src/Functions/array/arrayMinMaxIndex.cpp [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+        return lhs > rhs;
+}
+
+/// Keep the one-pass implementation for short arrays and types without a vectorized findExtreme implementation.
+template <ArrayMinMaxIndexStrategy strategy, typename T>
+static size_t findIndexOnePass(const T * data, size_t size)
+{
+    size_t index = 0;
+
+    if constexpr (std::is_floating_point_v<T>)
```
Comment (Algunenano) [inline src/Functions/array/arrayMinMaxIndex.cpp]: Why is this needed? Float32 and Float64 both have has_find_extreme_implementation. If we need BFloat16 (if supported here), let's add it there instead of duplicating all the NaN handling

File: src/Functions/array/arrayMinMaxIndex.cpp [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+    if constexpr (strategy == ArrayMinMaxIndexStrategy::Min)
+        return findExtremeMin(data, begin, end);
+    else
+        return findExtremeMax(data, begin, end);
+}
+
+template <typename T>
+static size_t findFirstExtreme(const T * data, size_t size, const T & extreme)
+{
+    if constexpr (std::is_floating_point_v<T>)
```
Comment (Algunenano) [inline src/Functions/array/arrayMinMaxIndex.cpp]: This shouldn't be needed. Handle it via existing functions

File: src/Functions/array/arrayMinMaxIndex.cpp [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+
+    const size_t index = findFirstExtreme(data, size, *extreme);
+    chassert(index < size);
+    return index;
+}
+
+template <ArrayMinMaxIndexStrategy strategy, typename T>
+static size_t findIndex(const T * data, size_t size)
+{
+    constexpr size_t min_size_for_extreme_reduction = 32;
```
Comment (Algunenano) [inline src/Functions/array/arrayMinMaxIndex.cpp]: Please, don't do this. I really dislike the fact that we keep adding random magic numbers to decide whether it's worth using a function or not, when performance it's not important and when that leads to duplicating the code and conditions (often times leading to bugs or to suboptimal code).

These kind of optimizations should only be done when it's clear they are needed. A new function that has 0 use does not need multiple paths to do the same thing

File: src/Functions/array/arrayMinMaxIndex.cpp (hunk 5 above)
Comment (Algunenano) [reply src/Functions/array/arrayMinMaxIndex.cpp]: Just checked. BFloat16 isn't supported. The problem is that you decided to duplicate code for a serial version under some magic constant

File: src/Functions/array/arrayMinMaxIndex.cpp [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+        {"source_arr", "The source array to process.", {"Array(T)"}},
+        {"cond1_arr, ...", "Optional. N condition arrays providing additional arguments to the lambda function.", {"Array(T)"}}
+    };
+    FunctionDocumentation::ReturnedValue returned_value_min = {"Returns the 1-based index of the first minimum element, or `0` for an empty array.", {"UInt32"}};
+    FunctionDocumentation::Examples examples_min = {
+        {"Basic example", "SELECT arrayMinIndex([5, 3, 2, 7]);", "3"},
+        {"First equal minimum", "SELECT arrayMinIndex([5, 3, 3, 7]);", "2"},
+        {"Usage with lambda function", "SELECT arrayMinIndex(x, y -> x/y, [4, 8, 12, 16], [1, 2, 1, 2]);", "1"},
+    };
+    FunctionDocumentation::IntroducedIn introduced_in_min = {26, 9};
```
Comment (Algunenano) [inline src/Functions/array/arrayMinMaxIndex.cpp]: Should be 26.10 now. Both places

File: tests/performance/array_min_max_index.xml [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+        FROM numbers(2000)</fill_query>
+    <fill_query>INSERT INTO array_min_max_index_perf_float
+        SELECT 2, arrayMap(x -&gt; toFloat64(x) - 8192, range(16384))
+        FROM numbers(40)</fill_query>
+
+    <fill_query>INSERT INTO array_min_max_index_perf_string
+        SELECT 1, arrayMap(x -&gt; concat('item-', toString(x)), range(32))
+        FROM numbers(10000)</fill_query>
+
+    <query>SELECT sum(arrayMinIndex(a) + arrayMaxIndex(a)) FROM array_min_max_index_perf_u64 WHERE case_id IN (1, 2) FORMAT Null</query>
```
Comment (Algunenano) [inline tests/performance/array_min_max_index.xml]: Please, use short and concise performance tests. If you need 10+ tests for a new function, you are overdoing it

File: tests/queries/0_stateless/05213_array_min_max_index.sql [hunk 10]
```diff
@@ -0,0 +1,194 @@
+SELECT arrayMinIndex([5, 3, 2, 7]), arrayMaxIndex([5, 3, 2, 7]);
```
Comment (Algunenano) [inline tests/queries/0_stateless/05213_array_min_max_index.sql]: I'm afraid, as is this is really hard to review. Consider explaining what some of the complex queries are doing and turning on echo (`-- { echoOn }`)

**Review by Algunenano — CHANGES_REQUESTED** [review]
This code is extremely complex for no need. When introducing new code, focus on it being correct and simple. It should be understandable at first glance. Consider tuning your LLM for that goal, rather than putting an extreme focus on performance (which is not the goal of this PR). Also, please consider readability. Otherwise these PRs won't be reviewed; human time is really limited.

File: src/Functions/array/arrayMinMaxIndex.cpp (hunk 4 above)
Comment (fallintoplace) [reply src/Functions/array/arrayMinMaxIndex.cpp]: They have different semantics on tie-breaks unfortunately.

Comment (Algunenano) [reply src/Functions/array/arrayMinMaxIndex.cpp]: Could you please explain? What semantics? What cases? Why can't the new functions have the same semantics?

Comment (fallintoplace) [reply src/Functions/array/arrayMinMaxIndex.cpp]: indexOf and indexOfAssumeSorted return the first match. For [5, 7, 7, 3], indexOf(arr, arrayMax(arr)) returns 2, while findExtremeMaxIndex picks the 3rd 7. For [7, 7, 5, 3], it picks the 1st 7.

arrayMaxIndex should use a predictable tie-breaker so callers know which index they’ll get. Returning the first match also preserves the result of the existing indexOf expression. In its unstable ORDER BY ... LIMIT 1 use, the helper doesn’t need to promise which tied row it picks. 

Same for min.
